### PR TITLE
Misc Changes

### PR DESCRIPTION
**Pull request by tarekgh**
The changes here include the following:

- Support `o1` model labels in the Tiktoken tokenizer
- Replace the usage of the Tuple in the `EncodedToken` with the `Range` and remove the `TorchSharp` `Range/Index` implementation
- Rename the `SentencePieceBpeTokenizer` to allow adding more models to it in the future
- Make the `Tokenizer.Decode` method return a non-nullable string
- Add support for added tokens in the BPE tokenizer

Every commit in the PR is representing one of the listed changes. Reviewing every commit separately will make it easier to have a clear understanding of the changes.

### CODE REVIEW

**Comment by tarekgh** [comment]
@michaelgsharp Could you please help review this change? 

@ericstj could you please help reviewing the changes in the csproj files for a new added `Microsoft.Bcl.Memory` dependency and enable restoring from net9 feed? The second commit include such changes.

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
     <!-- Remove once we have resolved the TorchSharp issue. -->
     <ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>None</ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>
+    <SuppressTfmSupportBuildWarnings>true</SuppressTfmSupportBuildWarnings>
```
Comment (ericstj) [inline docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:11]: Make sure to file an issue to remove all these when we remove `net6.0` from the repo.

**Review by ericstj — APPROVED** [review]
Dependency changes look ok, can you please review as well @michaelgsharp

File: eng/Versions.props [hunk 2]
```diff
@@ -41,6 +41,7 @@
     <GoogleProtobufVersion>3.27.1</GoogleProtobufVersion>
     <LightGBMVersion>3.3.5</LightGBMVersion>
     <MicrosoftBclHashCodeVersion>1.1.1</MicrosoftBclHashCodeVersion>
+    <MicrosoftBclMemoryVersion>9.0.0-rc.1.24431.7</MicrosoftBclMemoryVersion>
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply eng/Versions.props:44]: I opened https://github.com/dotnet/machinelearning/issues/7268 to tack this issue.

File: docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj (hunk 1 above)
Comment (tarekgh) [reply docs/samples/Microsoft.ML.AutoML.Samples/Microsoft.ML.AutoML.Samples.csproj:11]: I opened https://github.com/dotnet/machinelearning/issues/7268 to track this issue.

File: src/Microsoft.ML.Tokenizers/Model/EnglishRobertaTokenizer.cs [hunk 3]
```diff
@@ -597,14 +597,14 @@ private int EncodeToIdsResult(List<EncodedToken> tokens, IList<int>? accumulated
                 for (int i = 0; i < maxTokens; i++)
                 {
                     accumulatedIds.Add(tokens[i].Id);
-                    charsConsumed += tokens[i].Offset.Length;
+                    charsConsumed += tokens[i].Offset.End.Value - tokens[i].Offset.Start.Value;
```
Comment (michaelgsharp) [inline src/Microsoft.ML.Tokenizers/Model/EnglishRobertaTokenizer.cs:600]: Now that we are using `Range` and can use the "from the end" notation, is taht something we need to worry about here or is it being controlled/created in such a way it will never use that notation?

File: test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj [hunk 4]
```diff
@@ -5,6 +5,10 @@
     <ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>None</ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetFramework)' == 'net6.0'">
```
Comment (michaelgsharp) [inline test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj:8]: Do we want to create an issue to remember to drop this when we drop net6? Cause that will be happening soon.

File: src/Microsoft.ML.Tokenizers/Model/EnglishRobertaTokenizer.cs (hunk 3 above)
Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Model/EnglishRobertaTokenizer.cs:600]: Good question, we always fill the exact index as the length of the container will not be available all the time. It is a requirement for this scenario to fill the range with exact offsets.

File: src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs [hunk 5]
```diff
@@ -919,7 +919,7 @@ private Tensor PrepInputTensors(ref MLImage image, ValueGetter<MLImage> imageGet
                     var padW = 32 - (image.Width % 32);
                     var padH = 32 - (image.Height % 32);
                     var transMidTensor = torch.zeros(1, 3, image.Height + padH, image.Width + padW, device: _parent.Device);
-                    transMidTensor[.., .., ..image.Height, ..image.Width] = reMidTensor / 255.0;
+                    transMidTensor[RangeUtil.ToTensorIndex(..), RangeUtil.ToTensorIndex(..), RangeUtil.ToTensorIndex(..image.Height), RangeUtil.ToTensorIndex(..image.Width)] = reMidTensor / 255.0;
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs:922]: `transMidTensor` is accepting `TensorIndex`. And `..` is a Range type which is defined in another library now. So, need a way to convert `..` which is `Range` to `TensorIndex`. What can be done to make this syntax work again is to add implicit conversion inside the `TensorIndex` struct. @tannergooding may help advising about that?

[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs:922]: By the way, this syntax used to work before because we had our own implementation of `Range` which had implicit conversion to `TensorIndex`.

File: test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj (hunk 4 above)
Comment (tarekgh) [reply test/Microsoft.ML.AutoML.Tests/Microsoft.ML.AutoML.Tests.csproj:8]: I already opened https://github.com/dotnet/machinelearning/issues/7268 for tracking.

File: src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/Microsoft.ML.TorchSharp/AutoFormerV2/ObjectDetectionTrainer.cs:922]: @michaelgsharp let me know if you want open issue track that?
